forge/src/trace/decoder.rs

```diff
@@ -0,0 +1,234 @@
+    /// The call trace decoder always knows how to decode calls to the cheatcode address, as well
+    /// as DSTest-style logs.
+    pub fn new() -> Self {
+        Self {
+            contracts: BTreeMap::new(),
+            labels: [(*CHEATCODE_ADDRESS, "VM".to_string())].into(),
+            functions: HEVM_ABI
+                .functions()
+                .map(|func| (func.short_signature(), func.clone()))
+                .collect::<BTreeMap<[u8; 4], Function>>(),
+            events: CONSOLE_ABI
+                .events()
+                .map(|event| (event.signature(), event.clone()))
+                .collect::<BTreeMap<H256, Event>>(),
+            errors: Abi::default(),
+        }
+    }
+
+    /// Creates a new call trace decoder with predetermined address labels.
+    pub fn new_with_labels(labels: BTreeMap<Address, String>) -> Self {
+        let mut info = Self::new();
+        for (address, label) in labels.into_iter() {
+            info.labels.insert(address, label);
+        }
+        info
+    }
+
+    /// Identify unknown addresses in the specified call trace using the specified identifier.
+    ///
+    /// Unknown contracts are contracts that either lack a label or an ABI.
+    pub fn identify(&mut self, trace: &CallTraceArena, identifier: &impl TraceIdentifier) {
+        trace.addresses_iter().for_each(|(address, code)| {
+            // We only try to identify addresses with missing data
+            if self.labels.contains_key(address) && self.contracts.contains_key(address) {
+                return
+            }
+
+            let (contract, label, abi) = identifier.identify_address(address, code);
+            if let Some(contract) = contract {
+                self.contracts.entry(*address).or_insert(contract);
+            }
+
+            if let Some(label) = label {
+                self.labels.entry(*address).or_insert(label);
+            }
+
+            if let Some(abi) = abi {
+                // Flatten functions from all ABIs
+                abi.functions().map(|func| (func.short_signature(), func.clone())).for_each(
+                    |(sig, func)| {
+                        self.functions.insert(sig, func);
+                    },
+                );
+
+                // Flatten events from all ABIs
+                abi.events().map(|event| (event.signature(), event.clone())).for_each(
+                    |(sig, event)| {
+                        self.events.insert(sig, event);
+                    },
+                );
+
+                // Flatten errors from all ABIs
+                abi.errors().for_each(|error| {
+                    let entry = self
+                        .errors
+                        .errors
+                        .entry(error.name.clone())
+                        .or_insert_with(Default::default);
+                    entry.push(error.clone());
+                });
+            }
+        });
+    }
+
+    pub fn decode(&self, traces: &mut CallTraceArena) {
```

Comment:
All of the rightward drift here will be gone when (if) we get [if-let chains](https://rust-lang.github.io/rfcs/2497-if-let-chains.html). Same goes for some of the call trace display logic